tour_model_eval/label_processing.py

```diff
@@ -1,65 +1,77 @@
-                # to see which row will be converted
-                logging.debug("The following rows will be changed: %s", user_input_df.iloc[a])
-                user_input_df.iloc[a]["replaced_mode"] = user_input_df.iloc[a]['mode_confirm']
+    # change language first
+    user_input_df = map_labels_sp2en(user_input_df)
+    # convert purpose
+    user_input_df = user_input_df.replace(map_pur_dict)
+    # convert mode
+    for a in range(len(user_input_df)):
+        if user_input_df.iloc[a]["replaced_mode"] == "same_mode":
+            # to see which row will be converted
+            logging.debug("The following rows will be changed: %s", user_input_df.iloc[a])
+            user_input_df.iloc[a]["replaced_mode"] = user_input_df.iloc[a]['mode_confirm']
     return user_input_df
 
 
 # use hierarchical clustering to get labels of the second round
+# - sch.linkage: perform hierarchical(agglomerative) clustering
+# In this function, we set a low bound and a higher bound(cutoff) of distance in the dendrogram
+# - last_d: the distance of the last cluster in the dendrogram
+# - low: the lower bound of distance
+# e.g., if low = 300, last_d = 250, the list of clusters will be all 0s, like [0,0,0,0,0].
+# It means the points are already similar to each other after the first round of clustering, they don't need to
+# go through the second round.
+# - max_d: the cutoff of distance
+# - dist_pct: the percentage of the last distance in the dendrogram
+# - sch.fcluster: form clusters from the hierarchical clustering defined by the given linkage matrix
+# e.g., if last_d = 10000, dist_pct = 0.4, max_d = 400, clusters will be assigned at the distance of 400
 def get_second_labels(x,method,low,dist_pct):
-    z = linkage(x, method=method, metric='euclidean')
+    z = sch.linkage(x, method=method, metric='euclidean')
     last_d = z[-1][2]
     clusters = []
     if last_d < low:
         for i in range(len(x)):
             clusters.append(0)
     else:
         max_d = last_d * dist_pct
-        clusters = fcluster(z, max_d, criterion='distance')
+        clusters = sch.fcluster(z, max_d, criterion='distance')
     return clusters
 
 
 # this function includes hierarchical clustering and changing labels to get appropriate labels for
 # the second round of clustering
 # appropriate labels are label from the first round concatenate label from the second round
 # (e.g. label from first round is 1, label from second round is 2, the new label will be 12)
+# - second_round_idx_labels: a track to store the indices and labels for the second round
 def get_new_labels(x,low,dist_pct,second_round_idx_labels,new_labels,method=None):
-    idx_label = second_round_idx_labels.copy()
     second_labels = get_second_labels(x,method,low,dist_pct)
     for i in range(len(second_labels)):
-        index = idx_label[i][0]
-        new_label = idx_label[i][1]
+        index = second_round_idx_labels[i][0]
+        new_label = second_round_idx_labels[i][1]
```

Comment:
clarify what is the `index` and the `second_label`. If the `index` is the first label (which is what the code seems to show), then calling it `first_index` would be good 😄

Reply:
changed